packages/salesforcedx-vscode-apex-replay-debugger/src/telemetry/telemetry.ts

```diff
@@ -0,0 +1,56 @@
+import vscode = require('vscode');
+import TelemetryReporter from 'vscode-extension-telemetry';
+
+export class TelemetryService {
+  private static instance: TelemetryService;
+  private context: vscode.ExtensionContext | undefined;
+  private reporter: TelemetryReporter | undefined;
+
+  public static getInstance() {
+    if (!TelemetryService.instance) {
+      TelemetryService.instance = new TelemetryService();
+    }
+    return TelemetryService.instance;
+  }
+
+  public initializeService(
+    context: vscode.ExtensionContext,
+    isTelemetryEnabled: boolean
+  ): TelemetryReporter | undefined {
+    this.context = context;
+
+    // TelemetryReporter is not initialized if user has disabled telemetry setting.
+    if (this.reporter === undefined && isTelemetryEnabled) {
+      const extensionPackage = require(this.context.asAbsolutePath(
+        './package.json'
+      ));
+
+      this.reporter = new TelemetryReporter(
+        extensionPackage.name,
+        extensionPackage.version,
+        extensionPackage.aiKey
+      );
+      this.context.subscriptions.push(this.reporter);
+    }
+
+    return this.reporter;
+  }
+
+  public sendExtensionActivationEvent(): void {
```

Comment:
I want avoid having to create configuration data when using the Telemetry service across the extension so decided to create this methods. Let me know what you think of this approach.

Reply:
I think this is _safer_ and more useful. So, let's go with smaller and YAGNI first.

We might need to directly expose https://github.com/Microsoft/vscode-extension-telemetry/blob/master/src/telemetryReporter.ts#L88 in the future. But we can do that later as required. I'm just worried that we if we expose too much now, it would be hard to correct things.

In particular, we need to be careful of the keys, of `public sendTelemetryEvent(eventName: string, properties?: { [key: string]: string }, measures?: { [key: string]: number })`. If the caller of that screws it up, the data might not be easily processed.

By the way, how _powerful_ is AppInsights compared to splunk? Do they have good querying tools?